### ISSUE TITLE
controlnet tensorrt pipeline

### ISSUE THREAD

**Issue by hnsywangxin** [NVIDIA/TensorRT#2860]
Hi,I want to know,if I want to use stable diffusion with controlnet by tensorrt,What should I do，Can you give me some advice?
thanks!!!

**Comment by vanewu** [comment]
I did the conversion using depth's controlnet StableDiffusion. But the conversion method can be generalized to other types of controlnets.
Currently we do it in two ways:
1. Look this [code](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_img2img.py#L133 ), you will find that the combination of Controlnet and SD is mainly in [here](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_img2img.py#L133) . You can refer to the demo [Unet](https://github.com/NVIDIA/TensorRT/blob/release/8.6/demo/Diffusion/models.py#L261) to write a similar ControlnetUnet (You can refer to the samplecode below), and use a similar conversion method to convert.
```
class SampleControlUnet(nn.Module):
    def __init__(self, base_model_dir, control_model_dir, fp16=True) -> None:
        super().__init__()
        self.device = 'cuda:0'
        self.fp16 = fp16
        model_opts = {'revision': 'fp16',
                      'torch_dtype': torch.float16} if self.fp16 else {}
        self.base = UNet2DConditionModel.from_pretrained(
            base_model_dir, subfolder="unet",
            **model_opts
        ).eval().to(self.device)
        self.control = ControlNetModel.from_pretrained(
            control_model_dir,
            **model_opts
        ).eval().to(self.device)
    def forward(self, sample, timestep, encoder_hidden_states, controlnet_cond):
        controlnet_conditioning_scale: float = 1.0	# maybe set to hyperparams or input
        down_block_res_samples, mid_block_res_sample = self.control(
            sample,
            timestep,
            encoder_hidden_states=encoder_hidden_states,
            controlnet_cond=controlnet_cond,
            return_dict=False,
        )
        down_block_res_samples = [
            down_block_res_sample * controlnet_conditioning_scale
            for down_block_res_sample in down_block_res_samples
        ]
        mid_block_res_sample *= controlnet_conditioning_scale
        # predict the noise residual
        noise_pred = self.base(
            sample,
            timestep,
            encoder_hidden_states=encoder_hidden_states,
            down_block_additional_residuals=down_block_res_samples,
            mid_block_additional_residual=mid_block_res_sample,
        ).sample
        return noise_pred
```
2. Modify the input configuration of the demo [Unet](https://github.com/NVIDIA/TensorRT/blob/release/8.6/demo/Diffusion/models.py#L261), accept the output of controlnet (actually half of the unet structure), and set These outputs are passed as additional inputs to the Unet class in the demo.

The advantage of 1 is that the input and output are very concise,  the combination of unet and controlnet output can be completed automatically, and there is no need to manually save the input. But the disadvantage is that in this case controlnet and unet are bound together. Unable to do different controlnet and SD hotplug.
The advantage of 2 is that different controlnets can be combined with SD, because only the output of different controlnets needs to be sent to the converted Unet. But the disadvantage is that the input configuration of the entire Unet is not so concise.

The above two methods satisfy all my current applications. I have done a test and the generated effect is exactly the same as the diffusers. So I think the method is correct, I hope it can help you.

Maybe there is a better way, we can discuss it together.

**Comment by hnsywangxin** [comment]
@vanewu Thanks for your replay very much, text2img controlnet stable diffusion pipeline has been add in official repo, I want to use tensorRT to accelerate text2img using controlnet, What to look out for when I use tensorRT?
And, I use the demo, the tensorRT memory occupy is terrible(about 20G),It's strange.

Thank you again.

**Comment by YanBC** [comment]
Hi there, I too am looking for a way to accelate ControlNet with TensorRT. I don't see any codes related to the ControlNet pipeline you are talking about. Can you post a link? @hnsywangxin

**Comment by YanBC** [comment]
Are you guys working on a ControlNet demo? How's it going? @zerollzeng

**Comment by hnsywangxin** [comment]
@YanBC I am developing with controlnet, I haven't finish it

**Comment by zhangvia** [comment]
> @YanBC I am developing with controlnet, I haven't finish it

hey,after installing the tensorrt8.6,i fix my problem.and now i'm also working on controlnet with tensorrt. what should i do?overwrite the unet model? but i think models.py just for the onnx2trt covert progress. i have to write a new pipline? but how to write the pipline? any advices?

**Comment by ousinkou** [comment]
@vanewu I tried method 1, but exporting onnx was failed. The model size exceeds supported 2GB limit. How to solve it?

**Comment by hnsywangxin** [comment]
> @vanewu I tried method 1, but exporting onnx was failed. The model size exceeds supported 2GB limit. How to solve it?

you can refer to this repo:https://github.com/hnsywangxin/controlnet_stable_tensorrt

**Comment by hnsywangxin** [comment]
> > @YanBC I am developing with controlnet, I haven't finish it
> 
> hey,after installing the tensorrt8.6,i fix my problem.and now i'm also working on controlnet with tensorrt. what should i do?overwrite the unet model? but i think models.py just for the onnx2trt covert progress. i have to write a new pipline? but how to write the pipline? any advices?

I have finished it,you can refer to this repo:https://github.com/hnsywangxin/controlnet_stable_tensorrt

**Comment by ousinkou** [comment]
@hnsywangxin Thanks. After adding load_external_data option, problem solved.
`onnx.load(onnx_path, load_external_data=False)`

**Comment by keddyjin** [comment]
@hnsywangxin @YanBC @ousinkou @zhangvia Hi all, I've implemented the stable diffusion controlnet TensorRT inference pipeline, which supports multiple controlnets with dynamic batchsize and dynamic shape. [Repo](https://github.com/keddyjin/TensorRT_StableDiffusion_ControlNet/tree/keddyj/demoDiffusionControlnet/demo/Diffusion), with thanks to @hnsywangxin's work.

**Comment by hnsywangxin** [comment]
> @hnsywangxin @YanBC @ousinkou @zhangvia Hi all, I've implemented the stable diffusion controlnet TensorRT inference pipeline, which supports multiple controlnets with dynamic batchsize and dynamic shape. [Repo](https://github.com/keddyjin/TensorRT_StableDiffusion_ControlNet/tree/keddyj/demoDiffusionControlnet/demo/Diffusion), with thanks to @hnsywangxin's work.

nice

**Comment by dotieuthien** [comment]
I follow your code to convert onnx, I face with the problem of 2GB exceed of protobuf size, my solution is shape_inference.infer_shapes_path(unet_model_path, unet_model_path), and disable infer_shapes in Optimizer

**Comment by zhangvia** [comment]
> @hnsywangxin @YanBC @ousinkou @zhangvia Hi all, I've implemented the stable diffusion controlnet TensorRT inference pipeline, which supports multiple controlnets with dynamic batchsize and dynamic shape. [Repo](https://github.com/keddyjin/TensorRT_StableDiffusion_ControlNet/tree/keddyj/demoDiffusionControlnet/demo/Diffusion), with thanks to @hnsywangxin's work.

is it possible that we can dynamically load lora when using tensorrt? now if i have ten lora weights, i need to build engine for every lora combined with a base unet.what if i want to perserve one base model and ten loras, not ten base models with ten loras
